### DISCUSSION TITLE
rec v4模型训练，acc不提升，从第5个epoch开始一直到第500个始终维持在0.2

### DISCUSSION THREAD

**Discussion by cddddd6** [PaddlePaddle/PaddleOCR#13139] [Q&A]
#### 问题描述 / Problem Description
rec v4模型训练，acc和norm_edit_dis不提升，从第5个epoch开始一直到第500个，acc始终维持在0.2，norm_edit_dis维持在0.25，loss从1.6稳步下降至1.25
第五个epoch：
[2024/06/17 21:25:07] ppocr INFO: epoch: [5/500], global_step: 375, lr: 0.000500, acc: 0.162760, norm_edit_dis: 0.251667, CTCLoss: 0.419245, NRTRLoss: 1.264903, loss: 1.716696, avg_reader_cost: 0.00009 s, avg_batch_cost: 0.30337 s, avg_samples: 83.2, ips: 274.25515 samples/s, eta: 6:24:43, max_mem_reserved: 32275 MB, max_mem_allocated: 28751 MB
第490个epoch：
[2024/06/19 20:18:17] ppocr INFO: epoch: [490/500], global_step: 33810, lr: 0.000500, acc: 0.187500, norm_edit_dis: 0.310671, CTCLoss: 0.053128, NRTRLoss: 1.220689, loss: 1.272771, avg_reader_cost: 0.00020 s, avg_batch_cost: 0.59513 s, avg_samples: 156.8, ips: 263.47019 samples/s, eta: 0:07:08, max_mem_reserved: 29210 MB, max_mem_allocated: 27333 MB
[2024/06/19 20:18:19] ppocr INFO: cur metric, acc: 0.19544740819899994, norm_edit_dis: 0.2862414661770183, fps: 1205.2127416129172
best epoch：
[2024/06/19 20:18:19] ppocr INFO: best metric, acc: 0.20094191365037745, is_float16: False, norm_edit_dis: 0.22495511140089897, fps: 1212.9491793506809, best_epoch: 334

文本素材为长文本，文本示例如下：
H123_HNC_123KWH_3_WSS82_00237_V1.2.G.3_B_123789
H546_QSK_456Kwh_1_PQ_NGWS84_23469_V8.7.G.6_B_186238

用训练后的模型进行infer，发现大部分文本识别准确，但总会缺胳膊少腿，比如少一个字符，少一个.，或者少一个符号

#### 运行环境 / Runtime Environment
- OS:Linux
- Paddle:2.6.1
- PaddleOCR:2.7

#### config：ch_PP-OCRv4_rec.yml

```yml
Global:
  debug: false
  use_gpu: true
  epoch_num: 500
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output
  save_epoch_step: 10
  eval_batch_step: [0, 138]
  cal_metric_during_train: true
  pretrained_model: /PaddleOCR-main/rec_pretrained_model_v4/student.pdparams
  checkpoints:
  save_inference_dir: ./inference_model/
  use_visualdl: false
  infer_img: /PaddleOCR-main/infer_data
  character_dict_path: ppocr/utils/ppocr_keys_v1.txt
  max_text_length: &max_text_length 100
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ./infer/predicts_ppocrv4_pretrained_val.txt 


Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Const
    learning_rate: 0.0005
    warmup_epoch: 2
  regularizer:
    name: L2
    factor: 3.0e-05


Architecture:
  model_type: rec
  algorithm: SVTR_LCNet
  Transform:
  Backbone:
    name: PPLCNetV3
    scale: 0.95
  Head:
    name: MultiHead
    head_list:
      - CTCHead:
          Neck:
            name: svtr
            dims: 120
            depth: 2
            hidden_dims: 120
            kernel_size: [1, 3]
            use_guide: True
          Head:
            fc_decay: 0.00001
      - NRTRHead:
          nrtr_dim: 384
          max_text_length: *max_text_length

Loss:
  name: MultiLoss
  loss_config_list:
    - CTCLoss:
    - NRTRLoss:

PostProcess:  
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: MultiScaleDataSet
    ds_width: false
    data_dir: /PaddleOCR-main/train_data
    ext_op_transform_idx: 1
    label_file_list:
    - PaddleOCR-main/train_data/train.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - RecConAug:
        prob: 0.5
        ext_data_num: 2
        image_shape: [48, 320, 3]
        max_text_length: *max_text_length
    - RecAug:
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  sampler:
    name: MultiScaleSampler
    scales: [[320, 32], [320, 48], [320, 64]]
    first_bs: &bs 192
    fix_bs: false
    divided_factor: [8, 16] # w, h
    is_training: True
  loader:
    shuffle: true
    batch_size_per_card: *bs
    drop_last: true
    num_workers: 8
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: /PaddleOCR-main/train_data
    label_file_list:
    - /PaddleOCR-main/train_data/test.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - RecResizeImg:
        image_shape: [3, 48, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 128
    num_workers: 2
```

**Comment by SWHL** [comment]
数据集规模多大？字典多大？
建议提问尽量把问题描述清楚哈

**Reply by cddddd6** [reply]
数据集标注数据1000张，合成数据1w张，字典用的ppocr默认字典，6000多个字符

**Comment by SWHL** [comment]
我记得在ppocr技术报告中，ppocr训练文本识别数据是1700w左右吧。你这数据少了点

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年06月20日 11:54 |
| 收件人 | PaddlePaddle/PaddleOCR ***@***.***> |
| 抄送人 | SWHL ***@***.***>,
Comment ***@***.***> |
| 主题 | Re: [PaddlePaddle/PaddleOCR] rec v4模型训练，acc不提升，从第5个epoch开始一直到第500个始终维持在0.2 (Discussion #13139) |

数据集标注数据1000张，合成数据1w张，字典用的ppocr默认字典，6000多个字符

**Reply by cddddd6** [reply]
我用的v4预训练模型呀，我看官网说5000张就够了

**Reply by GreatV** [reply]
please refer to: https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_ch/training.md#33-%E8%87%AA%E5%B7%B1%E6%9E%84%E5%BB%BA%E6%95%B0%E6%8D%AE%E9%9B%86

**Reply by cddddd6** [reply]
> please refer to: https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_ch/training.md#33-%E8%87%AA%E5%B7%B1%E6%9E%84%E5%BB%BA%E6%95%B0%E6%8D%AE%E9%9B%86

我根据这篇文章的建议，把训练数据量提升到了20w，但acc始终维持在0.15， norm_edit_dis: 0.65左右，已经100多个epoch没有提升了，请大佬再帮忙看看
[2024/06/24 10:27:47] ppocr INFO: cur metric, acc: 0.14469147539147487, norm_edit_dis: 0.6432341714003962, fps: 1228.5428620471178
[2024/06/24 10:27:47] ppocr INFO: best metric, acc: 0.15023976532705594, is_float16: False, norm_edit_dis: 0.6529230221017059, fps: 1223.652338811894, best_epoch: 113
[2024/06/24 10:28:04] ppocr INFO: save model in ./output_recv4_bmu_rdb_20w/latest
[2024/06/24 10:28:08] ppocr INFO: epoch: [259/600], global_step: 374620, lr: 0.000614, acc: 0.148437, norm_edit_dis: 0.629676, CTCLoss: 0.067272, NRTRLoss: 1.272548, loss: 1.344786, avg_reader_cost: 1.80563 s, avg_batch_cost: 2.04161 s, avg_samples: 51.2, ips: 25.07819 samples/s, eta: 3 days, 9:05:17, max_mem_reserved: 31814 MB, max_mem_allocated: 28636 MB

**Reply by SWHL** [reply]
确定一下数据集在送入模型前，图像和标签是否对应，感觉这有些离谱。盲猜是数据集标签没对上。

**Reply by cddddd6** [reply]
> 确定一下数据集在送入模型前，图像和标签是否对应，感觉这有些离谱。盲猜是数据集标签没对上。

用代码检查了一下，文件夹内的图片和txt中的文件名都是一一对应的，随机抽查了10张图片，图片内容和txt中标签也是对应的

**Reply by SWHL** [reply]
送入训练前是否对应呢？或者可以搞一个很小的tiny数据集，看看模型能训的过拟合不

**Reply by cddddd6** [reply]
> 送入训练前是否对应呢？或者可以搞一个很小的tiny数据集，看看模型能训的过拟合不

请问下送入训练前是否对应是啥意思？我之前有过文件txt文件名忘改的的情况，模型本身也会报错说找不到路径，所以现在应该是对应的。另外我之前用500张图片训练过一次，acc能到达0.96，部分train log如下
[2024/06/14 10:59:51] ppocr INFO: cur metric, acc: 0.9545454003099204, norm_edit_dis: 0.9954005975061565, fps: 1069.8948710962184
[2024/06/14 10:59:51] ppocr INFO: best metric, acc: 0.960227218168908, is_float16: False, norm_edit_dis: 0.9952373602837797, fps: 1071.2238636165755, best_epoch: 234
[2024/06/14 10:59:54] ppocr INFO: epoch: [384/500], global_step: 2304, lr: 0.000067, acc: 0.955729, norm_edit_dis: 0.995420, CTCLoss: 0.258425, NRTRLoss: 1.206312, loss: 1.464732, avg_reader_cost: 0.00370 s, avg_batch_cost: 0.23348 s, avg_samples: 60.8, ips: 260.41269 samples/s, eta: 0:14:11, max_mem_reserved: 25435 MB, max_mem_allocated: 24887 MB

**Reply by SWHL** [reply]
我的意思是训练前，可能经过一系列前处理，标签错位了啥的。
500张能训出来，那说明整体没问题啊。
这就不好说了。也许你的任务太难了？

总感觉哪里有问题，这就需要你那里慢慢debug了

**Reply by cddddd6** [reply]
> 我的意思是训练前，可能经过一系列前处理，标签错位了啥的。 500张能训出来，那说明整体没问题啊。 这就不好说了。也许你的任务太难了？
> 
> 总感觉哪里有问题，这就需要你那里慢慢debug了
我就是用TextRecognitionDataGenerator随机生成的这种长字符串，包含字母，数字，特殊符号，应该不算难吧，字体用的是TimesNewRoman，示例如下：
bXZnOC1.LsV9YFeWSow3-M0fWLZixzeDCFA8XoEy.5X3wRw6Gc4BdT0ebKyi
4-W.-fO-igz0Aqx1eFJzn._UGyMkLt-mlc1DY2B4jF6X5GvCLbiP1FRV0O
aDr4XYIE2MlTqTP59OQLVu_MYyR.XNbnA4vNh0FYNHGIUy99qF8reYsqhQJU

**Reply by GreatV** [reply]
@cddddd6 , 老哥你这都是英文的符号，换成英文配置文件训练吧，中文比较难训练。

**Reply by cddddd6** [reply]
> @cddddd6 , 老哥你这都是英文的符号，换成英文配置文件训练吧，中文比较难训练。

老哥，我根据你的建议替换了配置文件和预训练模型为英文版，出现以下warning，请问是否会影响模型训练：
[2024/06/24 14:50:39] ppocr WARNING: The shape of model params head.ctc_head.fc.weight [120, 101] not matched with loaded params head.ctc_head.fc.weight [120, 97] !
[2024/06/24 14:50:39] ppocr WARNING: The shape of model params head.ctc_head.fc.bias [101] not matched with loaded params head.ctc_head.fc.bias [97] !
[2024/06/24 14:50:39] ppocr WARNING: The shape of model params head.gtc_head.embedding.embedding.weight [105, 384] not matched with loaded params head.gtc_head.embedding.embedding.weight [101, 384] !
[2024/06/24 14:50:39] ppocr WARNING: The shape of model params head.gtc_head.tgt_word_prj.weight [384, 105] not matched with loaded params head.gtc_head.tgt_word_prj.weight [384, 101] !

**Reply by SWHL** [reply]
我刚注意到你这用的是中文字典。正常情况是字典都是根据你的数据集而来的。数据集标签里有哪些字符，字典就是哪些。

换成英文配置的话，对应的预训练模型也要换成英文的。

**Reply by cddddd6** [reply]
> 我刚注意到你这用的是中文字典。正常情况是字典都是根据你的数据集而来的。数据集标签里有哪些字符，字典就是哪些。
> 
> 换成英文配置的话，对应的预训练模型也要换成英文的。

好的好的，已经换了在重新训练，我用之前那个中文rec模型evaluate了一下eval数据集，acc是0.15，但我用中文rec模型infer了一下eval数据集，发现实际准确率是0.28，请教下为啥这里acc和实际准确率会不一致

**Reply by SWHL** [reply]
这个自己debug一下吧，估计前后处理没统一

**Reply by cddddd6** [reply]
> 这个自己debug一下吧，估计前后处理没统一

大佬，我用预训练v4模型eval测试了一下，acc只有0.06，norm_edit_dis只有0.1。但我用模型infer了相同数据集10张图片，识别的准确率至少有90%，请问这个是啥原因，evaluate和infer时调用的参数会不一样吗？是否是因为识别长文本image_shape : [48, 320, 3]有问题？感谢！

**Reply by SWHL** [reply]
看你描述，我也不知道为啥哈。需要仔细debug，除模型外其他因素是否有不同的。

**Comment by GreatV** [comment]
检查一下数据集吧
